dbms/src/Storages/TrivialBuffer.cpp

```diff
@@ -0,0 +1,554 @@
+            throw Exception("Destination table is myself. Read will cause infinite loop.",
+                    ErrorCodes::INFINITE_LOOP);
+
+        /** Отключаем оптимизацию "перенос в PREWHERE",
+          *  так как Buffer не поддерживает PREWHERE.
+          */
+        Settings modified_settings = settings;
+        modified_settings.optimize_move_to_prewhere = false;
+
+        streams = destination->read(column_names, query, context, modified_settings,
+            processed_stage, max_block_size, threads);
+    }
+
+    streams.push_back(std::make_shared<TrivialBufferBlockInputStream>(column_names, *this));
+
+    /** Если источники из таблицы были обработаны до какой-то не начальной стадии выполнения запроса,
+      * то тогда источники из буферов надо тоже обернуть в конвейер обработки до той же стадии.
+      */
+    if (processed_stage > QueryProcessingStage::FetchColumns)
+        streams.back() = InterpreterSelectQuery(query, context, processed_stage, 0, streams.back()).execute().in;
+
+    return streams;
+}
+
+void TrivialBuffer::addBlock(const Block & block)
+{
+    SipHash hash;
+    block.updateHash(hash);
+    HashType block_hash = hash.get64();
+
+    std::lock_guard<std::mutex> lock(mutex);
+    if (current_hashes->find(block_hash) == current_hashes->end()
+        && previous_hashes->find(block_hash) == previous_hashes->end())
+    {
+        if (current_hashes->size() >= num_blocks_to_deduplicate / 2)
+        {
+            previous_hashes = std::move(current_hashes);
+            current_hashes = std::make_unique<DeduplicationBuffer>();
+        }
+        current_hashes->insert(block_hash);
+        current_rows += block.rows();
+        current_bytes += block.bytes();
+        data.push_back(block);
+
+        CurrentMetrics::add(CurrentMetrics::TrivialBufferRows, current_rows);
+        CurrentMetrics::add(CurrentMetrics::TrivialBufferBytes, current_bytes);
+    }
+    else
+    {
+        //NOTE: может быть, если нашли этот блок в previous,
+        // то надо его хэш перенести в current?
+        /*
+        auto it = previous_hashes->find(block_hash);
+        if (it != previous_hashes->end())
+        {
+            current_hashes->insert(it);
+            previous_hashes->erase(it);
+        }
+        */
+    }
+}
+
+void TrivialBuffer::flush(bool check_thresholds)
+{
+    Block block_to_write;
+    time_t current_time = time(0);
+
+    time_t time_passed = 0;
+
+    if (data.empty())
+        return;
+
+    {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        if (first_write_time)
+            time_passed = current_time - first_write_time;
+
+        if (check_thresholds)
+        {
+            if (!checkThresholdsImpl(current_rows, current_bytes, time_passed))
+                return;
+        }
+        else
+        {
+            if (current_rows == 0)
+                return;
+        }
+
+        /// Collecting BlockList into single block.
+        block_to_write = data.front().cloneEmpty();
```

Comment:
The problem is when a new block may be inserted into 'data' after merge but before 'data.clear()'. Is it possible?